logger/tests/integration_tests.rs

```diff
@@ -0,0 +1,395 @@
+        }))
+        .await
+        .unwrap()
+        .into_inner();
+
+    // Check that no more logs have been recorded.
+    assert_eq!(
+        response
+            .log_items
+            .into_iter()
+            .map(MinLogItem::from)
+            .collect::<Vec<_>>(),
+        expected
+    );
+}
+
+#[tokio::test]
+async fn generate_and_stream_logs() {
+    let port = pick_unused_port().unwrap();
+    let addr = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), port);
+    const DEPLOYMENT_ID: &str = "stream-logs-deployment-id";
+
+    // Start the logger server in the background.
+    tokio::spawn(async move {
+        let sqlite = Sqlite::new_in_memory().await;
+
+        Server::builder()
+            .layer(JwtScopesLayer::new(vec![Scope::Logs]))
+            .add_service(TraceServiceServer::new(ShuttleLogsOtlp::new(
+                sqlite.get_sender(),
+            )))
+            .add_service(LoggerServer::new(Service::new(sqlite.get_sender(), sqlite)))
+            .serve(addr)
+            .await
+            .unwrap()
+    });
+
+    // Ensure the server has started.
+    tokio::time::sleep(std::time::Duration::from_millis(300)).await;
+
+    // Start a subscriber and generate some logs.
+    generate_logs(port, DEPLOYMENT_ID.into(), foo, true);
+
+    // Connect to the logger server so we can fetch logs.
+    let dst = format!("http://localhost:{port}");
+    let mut client = LoggerClient::connect(dst).await.unwrap();
+
+    // Subscribe to stream
+    let mut response = client
+        .get_logs_stream(Request::new(LogsRequest {
+            deployment_id: DEPLOYMENT_ID.into(),
+        }))
+        .await
+        .unwrap()
+        .into_inner();
+
+    let log = timeout(std::time::Duration::from_millis(500), response.message())
+        .await
+        .unwrap()
+        .unwrap()
+        .unwrap();
+
+    assert_eq!(
+        MinLogItem::from(log),
+        MinLogItem {
+            level: LogLevel::Trace,
+            fields: json!({"message": "foo"}),
+        },
+    );
+
+    // Start a subscriber and generate some more logs.
+    generate_logs(port, DEPLOYMENT_ID.into(), bar, true);
+
+    let log = timeout(std::time::Duration::from_millis(500), response.message())
+        .await
+        .unwrap()
+        .unwrap()
+        .unwrap();
+
+    assert_eq!(
+        MinLogItem::from(log),
+        MinLogItem {
+            level: LogLevel::Trace,
+            fields: json!({"message": "bar"}),
+        },
+    );
+}
+
+/// Helper function to setup a tracing subscriber and run an instrumented fn to produce logs.
+/// The runtime flag determines if we add the deployment_id as an attribute in the tracer config,
+/// which the shuttle-runtime will do, this way all traces from runtime are recorded. If we set
+/// it to false, only spans instrumented with the deployment_id field will be captured.
+fn generate_logs(port: u16, deployment_id: String, generator: fn(String), runtime: bool) {
```

Comment:
Is there a way to refactor this into two functions so that you don't have to send down the `runtime` argument?

Reply:
Probably! :smile: This is a bit messy, I'll look into it.

Reply:
I refactored this into one function for runtime and one for services, they are both almost the same so they just call the old function with the runtime arg. Perhaps you had a better idea?

Reply:
About, I was thinking this function could take in a resources vec which those two functions can set rather than the `runtime` arg.

Reply:
Ah! I changed it to that now, and also added comments to the runtime/service functions.